weave/monitoring/openai.py

```diff
@@ -54,71 +54,112 @@ def openai_create_preprocess(span: monitor.SpanWithInputs) -> None:
+            existing_choices[index]["finish_reason"] = choice_update["finish_reason"]
+    return existing_choices
+
+
+def count_chat_completion_tokens(span: monitor.SpanWithInputsAndOutput, record: typing.Any):
+    summary = {}
+    encoding = tiktoken.encoding_for_model(record["model"])
+
+    prompt_tokens = (
+        encoding.encode(m["content"]) for m in span.inputs["messages"]
+    )
+    summary["prompt_tokens"] = sum(len(c) for c in prompt_tokens)
+
+    completion_tokens = (
+        encoding.encode(c["message"]["content"]) for c in record["choices"]
+    )
+    summary["completion_tokens"] = sum(len(c) for c in completion_tokens)
+    summary["total_tokens"] = (
+        summary["prompt_tokens"] + summary["completion_tokens"]
+    )
+    return summary
+
+
+def count_completion_tokens(span: monitor.SpanWithInputsAndOutput, record: typing.Any):
+    summary = {}
+    encoding = tiktoken.encoding_for_model(record["model"])
+
+    prompt_tokens = encoding.encode(span.inputs["prompt"])
+    summary["prompt_tokens"] = len(prompt_tokens)
+
+    completion_tokens = 0
+    for c in record["choices"]:
+        completion_tokens += len(encoding.encode(c["text"]))
+#    completion_tokens = (encoding.encode(c["text"] for c in record["choices"]))
+    summary["completion_tokens"] = completion_tokens
+    summary["total_tokens"] = (
+        summary["prompt_tokens"] + summary["completion_tokens"]
+    )
+    return summary
+
+def openai_create_postprocess(process_choice_fn: typing.Any, count_token_fn: typing.Any) -> typing.Any:
+    def post_process(span: monitor.SpanWithInputsAndOutput):
+        def wrapped_gen(gen: typing.Generator) -> typing.Generator:
+            # TODO: this needs to compute token usage.
+            record = None
+            for item in gen:
+                if record is None:
+                    record = {
+                        "id": item["id"],
+                        "object": item["object"],
+                        "created": item["created"],
+                        "model": item["model"],
+                        "choices": [],
+                    }
+                record["choices"] = process_choice_fn(record["choices"], item["choices"])
+                yield item
+            if record is not None:
+                span.summary.update(count_token_fn(span, record))
+
+            span.output = record
+            span.close()
+
+        if span.inputs.get("stream"):
+            return wrapped_gen(span.output)
+
+        # move usage to summary
+        usage = span.output.pop("usage")
+        for k, v in usage.items():
+            span.summary[k] = v
+
+        return span.output
+    return post_process
 
 
 mon = monitor.default_monitor()
 
-monitored_create = mon.trace(
-    preprocess=openai_create_preprocess, postprocess=openai_create_postprocess
-)(openai.ChatCompletion.create)
+monitored_create = lambda openai_func, process_choice_fn, count_token_fn: mon.trace(
```

Comment:
This would be cleaner to just be it's own function. Technically nothing wrong with this, but stylistically feels weird.

Reply:
sure thing, ill make this explicit rather than a lambda